### PR TITLE
Guide - Pre-Compile Sensu Plugins

### LINKED ISSUE #438: Guide - Pre-compiling native extensions for Sensu Plugins gem dependencies
Creating a copy/pasta of this guide for our docs (Thanks to @cwjohnston): https://gist.github.com/cwjohnston/159a2cb56c1b3ebfcba3d0777f218725

## This is an issue with:

- [ ] Bug (site functionality or styling)
- [ ] Errata (fix needed for doc content)
- [x] New content (guide wanted)

## Context
Pre-compiling native extensions for Sensu Plugins gem dependencies

### PR DESCRIPTION
**Pull request by Blue0ctober**
Closes #438 

## Description
A how-to guide on precompiling sensu-plugins for distribution.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New doc/guide
- [ ] Fixing errata
- [ ] Update (Add missing or refresh existing content)

## Checklist:
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] All tests have passed.

### CODE REVIEW

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Needs a `_` at the end to complete the block.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Needs a `_` at the end to complete the block.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: For some reason in local testing, this numbered list entry begins with a `1` and not a `2`. This is true for all the numbered list entries in this guide with multiple entries.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Do want `$` in our commands? I know @apaskulin mentioned not doing that in the 2.0 doc proposal. Or maybe the effort to remove them prior to 2.0 should come in another request. I don't think we should be adding more, though.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Do we want our commands to have `sudo` where needed? I know in other guides we do add them where they are needed. Seems we should to stay consistent.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Tried this and it makes me think the command isn't correct:

```
yum groupinstall -y "Development Tools"
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirrors.cat.pdx.edu
 * epel: mirrors.cat.pdx.edu
 * extras: mirror.riverfrontnetworks.com
 * updates: mirror.metrocast.net
Maybe run: yum groups mark install (see man yum)
No packages in any requested group available to install or update
```

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Not sure if checking for the gem prior to needs to be show. If it isn't added by installing sensu, you can presume that follows of the guide won't have it. The which gem after exporting the path feels like enough to me.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: nokogiri is in tickets near the start of the guide. Any reason it isn't here? Should unf_ext be as well?

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Maybe say something like `in a production/testing/development environment` instead  of `in real life`.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: I get a failure here:

```
[root@enterprise-server1 ~]# cd /tmp/gems/cache
[root@enterprise-server1 cache]# gem compile unf_ext-0.0.7.4.gem
Unpacking gem: 'unf_ext-0.0.7.4' in temporary directory...
ERROR:  While executing gem ... (Gem::InstallError)
    invalid gem: No such file or directory @ rb_sysopen - unf_ext-0.0.7.4.gem
```

**Comment by rgeniesse** [comment]
@Blue0ctober let me know if you want to pair on this and we can go over all my comments.

**Review by rgeniesse — DISMISSED** [review]
Github forced comment, see above comment.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (Blue0ctober) [reply content/sensu-core/0.29/guides/pre-compile-plugins.md]: Worked on Centos 7.3, what version are you running?

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Get an error here 

```gem compile unf_ext-0.0.7.4.gem
Unpacking gem: 'unf_ext-0.0.7.4' in temporary directory...
ERROR:  While executing gem ... (Gem::InstallError)
    invalid gem: No such file or directory @ rb_sysopen - unf_ext-0.0.7.4.gem```

@Blue0ctober we probably should do a pairing to speed this up. Let me know!

Edit: Disregard. I was running all my stuff as `root`, which isn't the correct mindset as the guide doesn't say to become `root`.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: I get a permission error here, but running as `sudo` it can't find `gem` as `sudo` isn't aware of the previous `export` command. Becoming root, it works. I am not sure how to resolve this.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: This needs a `sudo` on the `yum` command, as `&&` doesn't bring it with the previous `sudo tee` command.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (asachs01) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: @Blue0ctober I feel like since the convention in the rest of the docs is to use `{{< highlight $LANGUAGE>}}{{< /highlight >}}`, this should be reflected throughout the doc. I'll leave that to @apaskulin for the final say though.

**Review by asachs01 — CHANGES_REQUESTED** [review]

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (asachs01) [inline content/sensu-core/1.0/guides/pre-compile-plugins.md]: This command fails due to the gem being `unf_ext-0.0.7.5-x86_64-linux.gem`. I'd recommend modifying the instructions so that you're not going after a specific version, as this is going to cause a bit of headache as that version increments.

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (asachs01) [inline content/sensu-core/1.0/guides/pre-compile-plugins.md]: This also fails due to the version of nokogiri being `nokogiri-1.8.4.gem`. Let's not go after a specific version here.

Comment (asachs01) [reply content/sensu-core/1.0/guides/pre-compile-plugins.md]: Add to this that Ruby devel is required. I'm getting failures when trying to compile noki:
```
# gem compile nokogiri-1.8.1.gem --prune
Unpacking gem: 'nokogiri-1.8.1' in temporary directory...
ERROR:  While executing gem ... (Gem::InstallError)
    invalid gem: No such file or directory @ rb_sysopen - nokogiri-1.8.1.gem
[root@sensu-enterprise-test cache]# gem compile nokogiri-1.8.4.gem --prune
Unpacking gem: 'nokogiri-1.8.4' in temporary directory...
Building native extensions.  This could take a while...
ERROR:  While executing gem ... (Gem::Ext::BuildError)
    ERROR: Failed to build gem native extension.

    current directory: /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4/ext/nokogiri
/opt/sensu/embedded/bin/ruby -r ./siteconf20180911-22233-111auh8.rb extconf.rb
checking if the C compiler accepts ... yes
Building nokogiri using packaged libraries.
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/opt/sensu/embedded/bin/$(RUBY_BASE_NAME)
        --help
        --clean
        --use-system-libraries
/opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:308:in `to_specs': Could not find 'mini_portile2' (~> 2.3.0) among 260 total gem(s) (Gem::MissingSpecError)
Checked in 'GEM_PATH=/root/.gem/ruby/2.4.0:/opt/sensu/embedded/lib/ruby/gems/2.4.0', execute `gem env` for more information
        from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:320:in `to_spec'
        from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_gem.rb:65:in `gem'
        from extconf.rb:463:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4/lib/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4 for inspection.
Results logged to /tmp/d20180911-22233-o6z89h/nokogiri-1.8.4/lib/gem_make.out
```
This persists after installing ruby-devel via yum.

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (asachs01) [inline content/sensu-core/1.0/guides/pre-compile-plugins.md]: This is a bit of an odd ending. It just drops off without any call to action. I'd recommend an invitation to edit the doc, or submit an issue. You'll probably want to recap what you went over too.

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (asachs01) [inline content/sensu-core/1.0/guides/pre-compile-plugins.md]: I feel like we shouldn't leave a "possibly" in here. We should know without a doubt as to what the dependencies are and what needs to be present. The "possibly" here makes it come across that we don't know what we're talking about.

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (asachs01) [inline content/sensu-core/1.0/guides/pre-compile-plugins.md]: As a note, I didn't get to this point due to the errors encountered earlier.

**Review by asachs01 — CHANGES_REQUESTED** [review]

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Needs that space removed near the end so `NOTE` is properly handled.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: This needs `sudo` to work, but that causes the exported PATH to not be found. Can fix by giving the absolute path to `gem`, `/opt/sensu/embedded/bin/gem` to all `gem` references.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: I get this error:

```
Unpacking gem: 'nokogiri-1.8.4' in temporary directory...
Building native extensions.  This could take a while...
ERROR:  While executing gem ... (Gem::Ext::BuildError)
    ERROR: Failed to build gem native extension.

    current directory: /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4/ext/nokogiri
/opt/sensu/embedded/bin/ruby -r ./siteconf20180926-13682-61ubeq.rb extconf.rb
checking if the C compiler accepts ... yes
Building nokogiri using packaged libraries.
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/opt/sensu/embedded/bin/$(RUBY_BASE_NAME)
	--help
	--clean
	--use-system-libraries
/opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:308:in `to_specs': Could not find 'mini_portile2' (~> 2.3.0) among 52 total gem(s) (Gem::MissingSpecError)
Checked in 'GEM_PATH=/home/vagrant/.gem/ruby/2.4.0:/opt/sensu/embedded/lib/ruby/gems/2.4.0', execute `gem env` for more information
	from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/dependency.rb:320:in `to_spec'
	from /opt/sensu/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_gem.rb:65:in `gem'
	from extconf.rb:463:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4/lib/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4 for inspection.
Results logged to /tmp/d20180926-13682-1ip9vmd/nokogiri-1.8.4/lib/gem_make.out
```

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: Maybe do: "If Sensu upgrades Ruby, for example from 2.4 to 2.5, packages will need to be rebuilt."

And if you still like the current way, you'll at least need to change `upgrade` to `upgrades`.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: There is an trailing `"` here. Possible that was meant for line 49?

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [reply content/sensu-core/0.29/guides/pre-compile-plugins.md]: This is still an issue for me.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [reply content/sensu-core/0.29/guides/pre-compile-plugins.md]: This issue remains.

File: content/sensu-core/0.29/guides/pre-compile-plugins.md
Comment (rgeniesse) [inline content/sensu-core/0.29/guides/pre-compile-plugins.md]: I get an error here. I had to use the absolute path for `gem` though as `export` does not get to `sudo` (see comment above. That said, here is the error:

```
[vagrant@pluygintest cache]$ /opt/sensu/embedded/bin/gem compile nokogiri-1.8.4.gem --prune
Unpacking gem: 'nokogiri-1.8.4' in temporary directory...
ERROR:  While executing gem ... (Gem::InstallError)
    invalid gem: No such file or directory @ rb_sysopen - nokogiri-1.8.4.gem
```

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (Blue0ctober) [reply content/sensu-core/1.0/guides/pre-compile-plugins.md]: pinned the versions in the install, this should fix the above while still being verbose.

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (Blue0ctober) [reply content/sensu-core/1.0/guides/pre-compile-plugins.md]: added ruby devel and --use-system-libraries command, still shows a couple of warnings but builds successfully in my testing

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (Blue0ctober) [reply content/sensu-core/1.0/guides/pre-compile-plugins.md]: removed "possibly"

Comment (Blue0ctober) [reply content/sensu-core/1.0/guides/pre-compile-plugins.md]: added it to the install process

**Review by rgeniesse — APPROVED** [review]
🎉 🎉 🎉 🎉 🎉 🎉 .

Copy to all versions before merging, as I think `0.29` was the only one getting updates as we pushed through this.

File: content/sensu-core/1.5/guides/pre-compile-plugins.md
Comment (apaskulin) [inline content/sensu-core/1.5/guides/pre-compile-plugins.md]: We'll need to remove lines 7 and 8 and change the weight to `12` for all versions

File: content/sensu-core/1.0/guides/pre-compile-plugins.md
Comment (Blue0ctober) [reply content/sensu-core/1.0/guides/pre-compile-plugins.md]: updated with a nicer ending

File: content/sensu-core/1.5/guides/pre-compile-plugins.md
Comment (Blue0ctober) [reply content/sensu-core/1.5/guides/pre-compile-plugins.md]: removed and changed weight

**Pull request merged**
